Speed up diff

escapewindow opened the pull request:
In the firefoxci cluster, we're seeing diff runs of >5 minutes
regularly, and >15 min when there are many many changes. This puts a lot
of friction on rolling out urgent fixes: since Jenkins diffs, then
applies to stage, then diffs, then applies to firefoxci, an urgent
rollout can take upwards of 40 minutes where it took less than 5 before.

Fixes #192

escapewindow commented on tcadmin/diff.py:
```diff
@@ -17,6 +18,28 @@
+        ],
+        encoding="utf8",
+        stdout=subprocess.PIPE,
+    ).stdout
```
We may want to clean up the temp files here?

jwhitlock replied:
I'm writing a follow-up PR to clean up tempfiles

escapewindow commented on tcadmin/diff.py:
```diff
@@ -95,9 +118,7 @@ def contextualize(rangeInfo):
-    lines = patiencediff.unified_diff(
-        left, right, lineterm="", fromfile="current", tofile="generated", n=context
-    )
+    lines = fast_diff(left, right, context)
```
If we're concerned about supporting systems without a `diff`, we could either allow someone to fall back to `patiencediff` or allow for a commandline option to specify that. I'm not sure if there are any systems we want to support without a `diff` though.

petemoore reviewed (commented):
If we instead use [difflib](https://docs.python.org/3/library/difflib.html#module-difflib) from the standard library, would we avoid introducing binary dependencies?

escapewindow commented:
> If we instead use [difflib](https://docs.python.org/3/library/difflib.html#module-difflib) from the standard library, would we avoid introducing binary dependencies?

Yes, but I believe we wouldn't speed it up as significantly either.

petemoore commented:
Could we benchmark it to find out?

jwhitlock commented:
I think we can get close enough benchmark results for diff vs difflib like this:

```
export TASKCLUSTER_ROOT_URL=https://firefox-ci-tc.services.mozilla.com
changed_config=$(mktemp /tmp/tc-admin-changes.XXXXXX)
time tc-admin generate --without-secrets > $changed_config
current_config=$(mktemp /tmp/tc-admin-current.XXXXXX)
time tc-admin current --without-secrets > $current_config
time diff $current_config $changed_config
curl -q https://raw.githubusercontent.com/python/cpython/main/Tools/scripts/diff.py > diff.py
time python3 diff.py $current_config $changed_config
rm $current_config
rm $changed_config
```

The script [diff.py](https://raw.githubusercontent.com/python/cpython/main/Tools/scripts/diff.py) is an implementation of diff using difflib that is mentioned in the [difflib docs](https://docs.python.org/3/library/difflib.html#a-command-line-interface-to-difflib)

escapewindow commented:
> Could we benchmark it to find out?

Ok. We're currently on ci-config [`1c7f0384f9db8d54af8aeafce7f73c92b3dbedc3`](https://hg.mozilla.org/ci/ci-configuration/rev/1c7f0384f9db8d54af8aeafce7f73c92b3dbedc3). It appears that large changes take much longer to diff than small changes, so I'm benchmarking how long it'll run `ci-admin diff` without [`940dd4213e0046839ffae3f80cca4bce0e819aa2`](https://hg.mozilla.org/ci/ci-configuration/rev/940dd4213e0046839ffae3f80cca4bce0e819aa2) in it. The previous revision was [`a967c9f26c4d5658d1b9a48c6f89c92a2c39ed7d`](https://hg.mozilla.org/ci/ci-configuration/rev/a967c9f26c4d5658d1b9a48c6f89c92a2c39ed7d), so

```
# in ci-configuration
hg up -r a967c9f26c4d5658d1b9a48c6f89c92a2c39ed7d
```

I have my `tc-admin` clone added to my virtualenv via `python setup.py develop`, which means any local changes apply to my virtualenv. So when I have tc-admin set to 89549eba62ae662381c3b1eb755663f92310f61f (the tip of this PR), we get

```
# commandline diff
ci-admin diff —environment firefoxci  170.32s user 2.49s system 91% cpu 3:09.79 total
```

Then I updated tc-admin to main, or 041d7258539835dc46a1f5800bbbc46c048c66e5.

```
# patiencediff
ci-admin diff —environment firefoxci  1062.04s user 9.44s system 98% cpu 18:10.62 total
```

or 5-6x as long via patiencediff than vi acli.

Then I updated to https://github.com/escapewindow/tc-admin/tree/difflib-wip, which is Ben's wip from #192 but using [difflib instead of cli diff](https://github.com/escapewindow/tc-admin/commit/0cf54b7dc3c57a383b646d02f4cc1cbd4c94ce76):

```
# difflib
ci-admin diff --environment firefoxci  1057.13s user 9.14s system 97% cpu 18:08.79 total
```

Or pretty similar to patiencediff.

petemoore commented:
Thanks so much for checking that @escapewindow - then it certainly makes sense to use the diff executable - let's go with that! Many thanks!

escapewindow commented:
Thank you! Could we get a new release on [pypi](https://pypi.org/project/tc-admin/) ?

jwhitlock commented:
Let's get #196 in as well, then I can release.
